File: android/app/src/main/java/com/algorand/android/ui/rekey/RekeyConfirmationViewModel.kt
/*
 * Copyright 2022 Pera Wallet, LDA
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

package com.algorand.android.ui.rekey

import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.algorand.android.core.AccountManager
import com.algorand.android.core.BaseViewModel
import com.algorand.android.models.Account
import com.algorand.android.models.AccountCacheData
import com.algorand.android.models.SignedTransactionDetail
import com.algorand.android.repository.TransactionsRepository
import com.algorand.android.utils.AccountCacheManager
import com.algorand.android.utils.Event
import com.algorand.android.utils.MIN_FEE
import com.algorand.android.utils.Resource
import com.algorand.android.utils.analytics.logRekeyEvent
import com.google.firebase.analytics.FirebaseAnalytics
import kotlin.math.max
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class RekeyConfirmationViewModel @ViewModelInject constructor(
    private val firebaseAnalytics: FirebaseAnalytics,
    private val accountManager: AccountManager,
    private val accountCacheManager: AccountCacheManager,
    private val transactionsRepository: TransactionsRepository
) : BaseViewModel() {

    private var sendTransactionJob: Job? = null

    val feeLiveData = MutableLiveData<Long>()
    val transactionResourceLiveData = MutableLiveData<Event<Resource<Any>>>()

    init {
        getFee()
    }

    private fun getFee() {
        viewModelScope.launch(Dispatchers.IO) {
            transactionsRepository.getTransactionParams().use(
                onSuccess = { params ->
                    feeLiveData.postValue(max(REKEY_BYTE_ARRAY_SIZE * params.fee, params.minFee ?: MIN_FEE))
                }
            )
        }
    }

    fun sendRekeyTransaction(transactionDetail: SignedTransactionDetail.RekeyOperation) {
        if (sendTransactionJob?.isActive == true) {
            return
        }

        sendTransactionJob = viewModelScope.launch(Dispatchers.IO) {
            transactionsRepository.sendSignedTransaction(transactionDetail.signedTransactionData).use(
                onSuccess = {
                    with(transactionDetail) {
                        val newRekeyedAuthDetailMap = mutableMapOf<String, Account.Detail.Ledger>().apply {
                            if (accountCacheData.account.detail is Account.Detail.RekeyedAuth) {
                                putAll(accountCacheData.account.detail.rekeyedAuthDetail)
                            }
                            put(rekeyAdminAddress, ledgerDetail)
                        }
                        val authAccount = Account.create(
                            publicKey = accountCacheData.account.address,
                            detail = Account.Detail.RekeyedAuth.create(
                                accountCacheData.getAuthTypeAndDetail(),
                                newRekeyedAuthDetailMap
                            ),
                            accountName = accountCacheData.account.name
                        )
                        accountManager.addNewAccount(authAccount)
                    }
                    firebaseAnalytics.logRekeyEvent()
                    transactionResourceLiveData.postValue(Event(Resource.Success(Any())))
                },
                onFailed = { exception, _ ->
                    transactionResourceLiveData.postValue(Event(Resource.Error.Api(exception)))
                }
            )
        }
    }

    fun getAccountCacheData(address: String): AccountCacheData? {
        return accountCacheManager.accountCacheMap.value[address]
    }

    fun getCachedAccountName(address: String): String? {
        return accountCacheManager.accountCacheMap.value[address]?.account?.name
    }

    fun getCachedAccountAuthAddress(address: String): String? {
        return accountCacheManager.accountCacheMap.value[address]?.authAddress
    }

    fun getCachedAccountData(address: String): Account? {
        return accountCacheManager.accountCacheMap.value[address]?.account
    }

    companion object {
        private const val REKEY_BYTE_ARRAY_SIZE = 30
    }
}

File: ios/Classes/ViewControllers/Ledger/Rekey/Confirm/ViewModels/RekeyConfirmationViewModel.swift
// Copyright 2022 Pera Wallet, LDA

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//
//  RekeyConfirmationViewModel.swift

import Foundation
import UIKit

final class RekeyConfirmationViewModel {
    private(set) var oldImage: UIImage?
    private(set) var oldTransitionTitle: String?
    private(set) var oldTransitionValue: String?
    private(set) var newTransitionTitle: String?
    private(set) var newTransitionValue: String?
    private(set) var newLedgerImage: UIImage?
    private(set) var feeValue: String?
    
    init(
        account: Account,
        ledgerName: String?,
        newAuthAddress: String
    ) {
        bindOldImage(account)
        bindOldTransitionTitle(account)
        bindOldTransitionValue(account)
        bindNewTransitionTitle(ledgerName)
        bindNewTransitionValue(
            ledgerName: ledgerName,
            newAuthAddress: newAuthAddress
        )
        bindNewLedgerImage(account)
        bindFeeValue()
    }
}

extension RekeyConfirmationViewModel {
    private func bindOldImage(_ account: Account) {
        if account.requiresLedgerConnection() {
            oldImage = "ledger-gray".uiImage
        } else {
            oldImage = "standard-gray".uiImage
        }
    }

    private func bindOldTransitionTitle(_ account: Account) {
        if account.requiresLedgerConnection() {
            oldTransitionTitle = "ledger-rekey-ledger-old".localized
        } else {
            oldTransitionTitle = "ledger-rekey-ledger-passphrase".localized
        }
    }

    private func bindOldTransitionValue(_ account: Account) {
        if account.requiresLedgerConnection() {
            if let ledgerName = account.currentLedgerDetail?.name {
                oldTransitionValue = ledgerName
            } else {
                oldTransitionValue = account.name
            }
        } else {
            oldTransitionValue = "*************"
        }
    }

    private func bindNewTransitionTitle(_ ledgerName: String?) {
        if ledgerName != nil {
            newTransitionTitle = "ledger-rekey-ledger-new".localized
            return
        }

        newTransitionTitle = "wallet-connect-transaction-title-auth-address".localized
    }

    private func bindNewTransitionValue(
        ledgerName: String?,
        newAuthAddress: String
    ) {
        newTransitionValue = ledgerName.unwrap(or: newAuthAddress.shortAddressDisplay)
    }

    private func bindNewLedgerImage(_ account: Account) {
        newLedgerImage = account.typeImage
    }

    private func bindFeeValue() {
        let fee = max(UInt64(minimumFee), 0)
        feeValue = "ledger-rekey-total-fee".localized(params: "\(fee.toAlgos)")
    }
}
